grease-cli/tests/Main.hs

```diff
@@ -460,28 +223,81 @@ withCapturedLogs withLogAction = do
   let logTxt = Text.unlines (map (Text.pack . show . PP.pretty) logs)
   pure logTxt
 
--- | Make an "Oughta"-based test for an S-expression or LLVM bitcode program
+-- | Make an "Oughta"-based test
 oughta ::
-  (SimOpts -> GreaseLogAction -> IO Results) ->
+  (GreaseLogAction -> IO Results) ->
   FilePath ->
   Oughta.LuaProgram ->
   IO ()
 oughta go path prog0 = do
-  opts <- getNonExeTestOpts path []
+  let action =
+        withCapturedLogs $ \la' -> do
+          res <- go la'
+          logResults la' res
   logTxt <-
-    withCapturedLogs $ \la' -> do
-      res <- go opts la'
-      logResults la' res
+    X.try @X.SomeException action <&>
+      \case
+        Left e -> Text.pack ("Exception: " ++ show e)
+        Right t -> t
   Text.IO.writeFile (FilePath.replaceExtension path "out") logTxt
   let output = Oughta.Output (Text.encodeUtf8 logTxt)
   let prog = Oughta.addPrefix prelude prog0
   Oughta.Result r <- Oughta.check prog output
   case r of
-    Left f -> throwIO f
+    Left f -> X.throwIO f
     Right s ->
       let ms = Oughta.successMatches s in
       T.U.assertBool "Test has some assertions" (not (Seq.null ms))
 
+-- | Make a "Oughta"-based tests for binaries
+oughtaBin ::
+  (GreaseLogAction -> IO Results) ->
+  Arch ->
+  -- | Directory
+  FilePath ->
+  -- | File
+  FilePath ->
+  T.TestTree
+oughtaBin go arch dir fileName =
+  T.U.testCase (FilePath.takeBaseName dir) $ do
+    let drop1 = FilePath.dropExtension fileName
+    let drop2 = FilePath.dropExtension drop1
+    let path = dir </> FilePath.addExtension drop2 "c"
+    content <- Text.IO.readFile path
+    let isArchComment =
+          Text.stripPrefix $
+            case arch of
+              Armv7 -> "// arm: "
+              PPC32 -> "// ppc32: "
+              X64 -> "// x64: "
+    let isGenericComment = Text.stripPrefix "// all: "
```

Comment:
Do you think we should use triple slashes (`///`) instead of double (`//`) to match the existing conventions in `tests/llvm-bc`?

Reply:
I don't have a strong opinion. The triple slashes in the bitcode tests were to disambiguate Oughta comments from other comments, but that isn't necessary here because we have these additional prefixes that name different architectures. I opted for maximal brevity rather than maximal consistency, I think either is reasonable.

Reply:
Fair enough. I don't have a strong opinion either, so I'm content to use the existing conventions.